internal/client/db/db.go

```diff
@@ -0,0 +1,585 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/internal/client/api"
+	"github.com/ChainSafe/gossamer/internal/client/db/columns"
+	"github.com/ChainSafe/gossamer/internal/client/db/metakeys"
+	"github.com/ChainSafe/gossamer/internal/primitives/blockchain"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/database"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"github.com/li1234yun/gods-generic/maps/linkedhashmap"
+)
+
+const numCachedHeaders = 8
+
+// An extrinsic entry in the database.
+type dbExtrinsic[E runtime.Extrinsic] struct {
+	inner any
+}
+
+type dbExtrinsicValues[E runtime.Extrinsic] interface {
+	dbExtrinsicIndexed | dbExtrinsicFull[E]
+}
+
+func setVote[E runtime.Extrinsic, Value dbExtrinsicValues[E]](mvdt *dbExtrinsic[E], value Value) {
+	mvdt.inner = value
+}
+
+func (mvdt *dbExtrinsic[E]) SetValue(value any) (err error) {
+	switch value := value.(type) {
+	case dbExtrinsicIndexed:
+		setVote[E](mvdt, value)
+		return
+	case dbExtrinsicFull[E]:
+		setVote[E](mvdt, value)
+		return
+	default:
+		return fmt.Errorf("unsupported type")
+	}
+}
+
+func (mvdt dbExtrinsic[E]) IndexValue() (index uint, value any, err error) {
+	switch mvdt.inner.(type) {
+	case dbExtrinsicIndexed:
+		return 0, mvdt.inner, nil
+	case dbExtrinsicFull[E]:
+		return 1, mvdt.inner, nil
+	}
+	return 0, nil, scale.ErrUnsupportedVaryingDataTypeValue
+}
+
+func (mvdt dbExtrinsic[E]) Value() (value any, err error) {
+	_, value, err = mvdt.IndexValue()
+	return
+}
+
+func (mvdt dbExtrinsic[E]) ValueAt(index uint) (value any, err error) {
+	switch index {
+	case 0:
+		return *new(dbExtrinsicIndexed), nil
+	case 1:
+		return *new(dbExtrinsicFull[E]), nil
+	}
+	return nil, scale.ErrUnsupportedVaryingDataTypeValue
+}
+
+// Extrinsic that contains indexed data.
+type dbExtrinsicIndexed struct {
+	// Hash of the indexed part.
+	Hash hash.H256
+	// Extrinsic header.
+	Header []byte
+}
+type dbExtrinsicFull[E runtime.Extrinsic] struct {
+	Extrinsic E
+}
+
+type metaUpdate[H, N any] struct {
+	Hash        H
+	Number      N
+	IsBest      bool
+	IsFinalized bool
+	WithState   bool
+}
+
+// blockchainDB is the block database.
+type blockchainDB[H runtime.Hash, N runtime.Number, E runtime.Extrinsic, Header runtime.Header[N, H]] struct {
+	db                   database.Database[hash.H256]
+	meta                 meta[H, N]
+	metaMtx              sync.RWMutex
+	leaves               api.LeafSet[H, N]
+	leavesMtx            sync.RWMutex
+	headerMetadataCache  blockchain.HeaderMetadataCache[H, N]
+	headerCache          linkedhashmap.Map[H, *runtime.Header[N, H]]
+	headerCacheMtx       sync.Mutex
+	pinnedBlocksCache    pinnedBlocksCache[H]
```

Comment:
I see that in the PR description the `db.pinnedBlocksCache` is defined as:

```
Manages pinning of blocks in a fixed size LRU cache. Entries are reference counted along with their justifications and block body.
```

So, what will be the use case for it? Also `blockchain.HeaderMetadataCache` is some sort of caching, can we group them in a single caching instance using the header hash as key?

Reply:
I just watched the part you explain about pinned blocks (when prune is enabled). So my question remains, can we have a single caching structure that holds those fast access blocks and in them the entry of this caching structure holds the block info + `isPinned` to avoid a unexpected pruning?

Reply:
`blockchain.HeaderMetadataCache` is only used in the methods that fulfill `blockchain.HeaderMetadata` for `blockchainDB`.  So this doesn't really have to do with pinning, but rather just caching the `HeaderMetadata` to avoid loading from the underlying DB every time, so I think it makes sense to be separate from the `pinnedBlocksCache`.  It's also used by other substrate code which is why it's in `blockchain` package.  But I will revisit in future work to see if it needs to be in a separate package.

Whether or not a block is pinned is checked in the pruning logic which will be implemented in a future PR.  It is separate in the fact that the `db.Backend` will explicitly check if something is pinned before generating the db commit to actually prune from the underlying DB.  In a future PR I will introduce an analogous `BlockImportOperation`, which acts on a parent state.  The parent block is pinned at this point to ensure we don't prune the parent block in the event of finalization.   I think I'll have a better idea on how to refactor once I actually implement the pruning.  But this workflow of pinning a block before working on top of it is prevalent not only for `blockchainDB` but also `StateDB` which is already on this feature branch.